betty/gramps/loader.py

```diff
@@ -1175,3 +1242,432 @@ def _load_attributes_for(
             self._load_attribute_privacy(entity, element, tag)
         if isinstance(entity, HasLinks):
             self._load_attribute_links(entity, gramps_entity_reference, element, tag)
+
+
+@internal
+class GrampsApiLoader(_GrampsLoader):
+    """
+    Load Gramps family history data into a project.
+    """
+
+    @override
+    async def load_xml(self, xml: str) -> None:
+        async with NamedTemporaryFile(mode="w", suffix=".gramps") as temporary_f:
+            with gzip.open(
+                temporary_f.name,  # type: ignore[arg-type]
+                "w",
+            ) as gzip_f:
+                gzip_f.write(
+                    xml.encode("utf-8")  # type: ignore[arg-type]
+                )
+            return await self.load_file(
+                Path(
+                    temporary_f.name  # type: ignore[arg-type]
+                )
+            )
+
+    @override
+    async def load_file(self, file_path: Path) -> None:
+        async with TemporaryDirectory() as temporary_directory_path_str:
+            database: DBAPI = make_database("sqlite")
+            database.load(":memory:")
+            database.set_mediapath(temporary_directory_path_str)
+            database_state = DbState()
+            database_state.db = database
+            user = User()
+            self._import_file(file_path, database_state, user)
+            await self._load_database(database)
+
+    def _import_file(
+        self, file_path: Path, database_state: DbState, user: User
+    ) -> None:
+        # Check proactively, because Gramps' importers do not raise exceptions for many things that can go wrong.
+        if not file_path.exists():
+            raise GrampsFileNotFound.new(file_path)
+        family_tree_format = file_path.suffix[1:]
+        plugin_manager = BasePluginManager.get_instance()
+        supported_extensions = []
+        for plugin in plugin_manager.get_import_plugins():
+            if plugin.get_extension() == family_tree_format:
+                import_function = plugin.get_import_function()
+                import_function(database_state.db, file_path, user)
+                return
+            supported_extensions.append(plugin.get_extension())
+        raise UserFacingGrampsError(
+            _(
+                "Gramps cannot import {file_path}. Supported file formats are {supported_extensions}"
+            ).format(
+                file_path=str(file_path),
+                supported_extensions=", ".join(supported_extensions),
+            )
+        )
+
+    async def _load_database(self, database: DBAPI) -> None:
```

Comment:
Make this method and anything it calls synchronous. Gramps' database isn't asynchronous anyway and we have no other IO

Reply:
Actually we call asynchronous factories so we have to keep things async.